### PR TITLE
Add kaiser_bessel_derived_window #17978 to tf_frontend_singal

### PR DESCRIPTION
**Pull request by Enigmafiend**
Close https://github.com/unifyai/ivy/issues/17978

### CODE REVIEW

**Comment by Enigmafiend** [comment]
ivy-gardener

**Comment by Enigmafiend** [comment]
add_frontend_checklist

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py [hunk 1]
```diff
@@ -91,3 +91,50 @@ def test_tensorflow_idct(
         norm=norm,
         atol=1e-01,
     )
+
+
+# kaiser_bessel_derived_window
+@handle_frontend_test(
+    fn_tree="tensorflow.signal.kaiser_bessel_derived_window",
+    dtype_and_window_length=helpers.dtype_and_values(
+        available_dtypes=(["int32"]),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:100]: Can you use `helpers.get_dtypes("valid")` instead?

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py [hunk 2]
```diff
@@ -91,3 +91,50 @@ def test_tensorflow_idct(
...
+        min_dim_size=1,
+        min_value=3,
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:102]: Is that necessary? Limiting length to 20 makes sense but I don't see any reason to limit the min value to a value bigger than 0.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py [hunk 3]
```diff
@@ -91,3 +91,50 @@ def test_tensorflow_idct(
...
+        min_num_dims=1,
+        max_value=20,
+        max_dim_size=1,
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:105]: Use of `shape` instead of `dim_size` would be better

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py [hunk 4]
```diff
@@ -91,3 +91,50 @@ def test_tensorflow_idct(
...
+        max_num_dims=1,
+    ),
+    dtype_and_beta=helpers.dtype_and_values(
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:108]: I think using `st.floats` and sampling float types would be better?

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py [hunk 5]
```diff
@@ -91,3 +91,50 @@ def test_tensorflow_idct(
...
+    helpers.test_frontend_function(
+        input_dtypes=[window_length_dtype[0], beta_dtype[0]],
+        frontend=frontend,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        window_length=window_length[0][0],
+        beta=beta[0][0],
+        rtol=1e-03,
+        atol=1e-03,
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:139]: I think those are not necessary

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @Enigmafiend 

I've made some comments that need to be addressed before merging your PR. Please take a look at those and to the updated checklist as well

Thank you so much for contributing to ivy 🚀🚀

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by Enigmafiend** [comment]
ivy-gardener

**Comment by Enigmafiend** [comment]
Fixed the test case problems and while the rtol and atol arent needed I add them in signals related tests for tradition sake, quantization errors and such from an engineering perspective do demand such from time to time

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py (hunk 2 above)
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py]: A 0 length window does not exist and is a trivial input (It will return an error or an empty array depending on the backend) as the number of number of elements for a kaiser window are always >0 and this is dictated by the window_length

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py [hunk 6]
```diff
@@ -91,3 +84,32 @@ def test_tensorflow_idct(
         norm=norm,
         atol=1e-01,
     )
+
+
+# kaiser_window
+@handle_frontend_test(
+    fn_tree="tensorflow.signal.kaiser_bessel_derived_window",
+    window_length=helpers.ints(min_value=1, max_value=100),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:92]: As per [tensorflow docs](https://www.tensorflow.org/api_docs/python/tf/signal/kaiser_bessel_derived_window), this should be a scalar **Tensor**. Please use a tensor with shape `(1,)` instead

Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:92]: Also we should test edge cases, so I suggest we add 0 as well

**Review by KareemMAX — CHANGES_REQUESTED** [review]
I've pointed some comments I have, please look into the test function and be sure to add `dtype` as a parameter to test on. We should always test the full input space.

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py (hunk 6 above)
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:92]: > tensor with shape `(1,)`

This strategy does produce a the correct tensor, the kaiser window also requires a rank of 0
image
image
image

Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:92]: > Also we should test edge cases, so I suggest we add 0 as well

Correct me if im wrong but 0 is a completely invalid input and will result in an InvalidArgument error as division by 0s is simply not possible and the Kaiser Window equation dictates a division by window length and thus is not supposed to be included in the test
image

Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:92]: > 
image

Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:92]: It should be noted that I use outputs of kaiser window here instead of the bessel function because at the backend both of them employ the tf.signal.kaiser_window func and just that the bessel derived performs a cumsum and sqr ontop, the core still remains the kaiser window

Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py]: > > Also we should test edge cases, so I suggest we add 0 as well
> 
> Correct me if im wrong but 0 is a completely invalid input and will result in an InvalidArgument error as division by 0s is simply not possible and the Kaiser Window equation dictates a division by window length and thus is not supposed to be included in the test image

I believe tensorflow doesn't agree with you on that.
```python
>>> tf.signal.kaiser_bessel_derived_window(tf.constant(0))
<tf.Tensor: shape=(0,), dtype=float32, numpy=array([], dtype=float32)>
```

Comment (KareemMAX) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py]: > > tensor with shape `(1,)`
> 
> This strategy does produce a the correct tensor, the kaiser window also requires a rank of 0 image image image

My bad, it should be `shape=(,)` instead 😅. I don't really like the use of `ints` as they don't try different dtypes. I'm asking internally about that.

File: ivy/functional/frontends/tensorflow/signal.py [hunk 7]
```diff
@@ -19,3 +19,14 @@ def kaiser_window(window_length, beta=12.0, dtype=ivy.float32, name=None):
...
+
+@handle_tf_dtype
+@to_ivy_arrays_and_back
+def kaiser_bessel_derived_window(
+    window_length, beta=12.0, dtype=ivy.float32, name=None
+):
+    return ivy.kaiser_bessel_derived_window(window_length, beta=beta, dtype=dtype)
+
+
+kaiser_window.supported_dtypes = ("float32", "float64", "float16", "bfloat16")
```
Comment (KareemMAX) [inline ivy/functional/frontends/tensorflow/signal.py:32]: Can you use `@with_unsupported_dtype ` instead?

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py [hunk 8]
```diff
@@ -91,3 +84,32 @@ def test_tensorflow_idct(
         norm=norm,
         atol=1e-01,
     )
+
+
+# kaiser_bessel_derived_window
+@handle_frontend_test(
+    fn_tree="tensorflow.signal.kaiser_bessel_derived_window",
+    window_length=helpers.ints(min_value=0, max_value=100),
+    beta=helpers.floats(min_value=0.0, max_value=80),
```
[the remark this answers is not in the document]
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:93]: I think i mentioned this but apart from TF most models dont support large betas due to overflow, esp jax and numpy and will crash with errors such as 
image

**Comment by Enigmafiend** [comment]
> Hey @Enigmafiend,
> 
> I've talked with @AnnaTz to make sure that I'm getting the details of the `dtype` issue right.
> 
> 1. Please don't add any changes unrelated to `kaiser_bessel_derived_window`.
> 2. You still need to add `dtype` to your test. It's required to test for every testable parameter.
> 3. You should generate `window_length` by using `dtype_and_values` with `helpers.get_dtypes("valid")`. You'll need to change the supported dtypes to integer ones.
> 4. Please don't add limits other than `min_value=0`
> 5. `beta` can be left being generated by `helpers.floats` but there is no need for limiting it

I havent made any changes unrelated to the current function, its just that for it to work for paddle, my kaiser_window fix will have to go through. Secondly 
 _3. You should generate `window_length` by using `dtype_and_values` with `helpers.get_dtypes("valid")`. You'll need to change the supported dtypes to integer ones._  will also result in tensors with ranks > 0 and hence will provide invalid inputs 
 
 Third _> 4. Please don't add limits other than `min_value=0`_ will cause the test to crash with extremely large inputs and as Ive mentioned in DMs and here aswell large betas above 80 will cause overflow aswell (Photo attached in the above thread)
image

I mean no offense but does the person you asked this about know why and how a kaiser window is used?

**Comment by Enigmafiend** [comment]
As for Beta, apart from TF which normalizes overflow, other models go upto approx 88.7 for beta
image
after which they overflow as again betas above said value are not practically used or exist (I showed the math for this in DMs)
image

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:93]: > There is no needs for those limits, having them will make our test partial
> 
> ```python
> >>> tf.signal.kaiser_bessel_derived_window(tf.constant(100), beta=1000)
> <tf.Tensor: shape=(100,), dtype=float32, numpy=
> array([0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00,
>        0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00,
>        0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00,
>        0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 5.4010830e-19,
>        1.9944142e-15, 2.7494018e-12, 1.4792860e-09, 3.2221595e-07,
>        2.9274252e-05, 1.1364573e-03, 1.9227374e-02, 1.4446096e-01,
>        4.9765486e-01, 8.6737514e-01, 9.8951054e-01, 9.9981517e-01,
>        9.9999934e-01, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 1.0000000e+00,
>        1.0000000e+00, 1.0000000e+00, 1.0000000e+00, 9.9999934e-01,
>        9.9981517e-01, 9.8951054e-01, 8.6737514e-01, 4.9765486e-01,
>        1.4446096e-01, 1.9227374e-02, 1.1364573e-03, 2.9274252e-05,
>        3.2221595e-07, 1.4792860e-09, 2.7494018e-12, 1.9944142e-15,
>        5.4010830e-19, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00,
>        0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00,
>        0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00,
>        0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00],
>       dtype=float32)>
> >>> tf.signal.kaiser_bessel_derived_window(tf.constant(100), beta=-1000)
> <tf.Tensor: shape=(100,), dtype=float32, numpy=
> array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
>        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
>        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
>        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
>        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
>        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
>        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
>        nan, nan, nan, nan, nan, nan, nan, nan, nan], dtype=float32)>
> ```

I will also point out that the negative beta array arent real KBDWs as their shapes range so
image

[the remark this answers is not in the document]
Comment (Enigmafiend) [reply ivy_tests/test_ivy/test_frontends/test_tensorflow/test_signal.py:93]: what TF is probably doing is recognizing and catching the overflow before it happens and giving the largest Beta input corresponding output instead (The Pink Rect signal). If you so want we could include such a modification to other backends as they already have a similar modification for window_length <2

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Please address the changes I mentioned in our last review. While keeping `window_length` to have a minimum value of 0 and no upper bound set. And beta can be bounded with `[0, 85]`

**Comment by Enigmafiend** [comment]
Hey, fixed an issue with Kaiser Window that would translate here aswell, as soon as that PR passes Ill update this one too

**Comment by KareemMAX** [comment]
> Hey, fixed an issue with Kaiser Window that would translate here aswell, as soon as that PR passes Ill update this one too

Cool, waiting for you
